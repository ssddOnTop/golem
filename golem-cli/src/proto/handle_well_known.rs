use anyhow::anyhow;
use tailcall_valid::{Valid, Validator};
use crate::proto::DEFAULT_INTERFACE_NAME;
use crate::wit_config::config::{Interface, WitConfig};
use crate::wit_config::wit_types::well_known::grpc_errors;

#[derive(derive_setters::Setters, Default)]
pub struct AppendedTypes {
    pub appended_error: String,
}

pub fn handle_well_known(config: WitConfig) -> Valid<(WitConfig, AppendedTypes), anyhow::Error, anyhow::Error> {
    Valid::succeed(config)
        .and_then(|mut config| {
            Valid::from_option(config.interfaces.iter().find(|v| v.name == DEFAULT_INTERFACE_NAME), anyhow!("Expected default interface to be present"))
                .map(|v| (*v).clone())
                .and_then(append_grpc_err)
                .map(|(interface, err_ty)| {
                    config.interfaces.remove(&interface);
                    config.interfaces.insert(interface);
                    (config, AppendedTypes::default().appended_error(err_ty))
                })
        })
}

fn append_grpc_err(interface: Interface) -> Valid<(Interface, String), anyhow::Error, anyhow::Error> {
    let mut error_ty = "grpc-errors".to_string();
    Valid::succeed(interface)
        .and_then(|mut interface| {
            if !interface.varients.contains_key(&error_ty) {
                interface.varients.insert(error_ty.clone(), grpc_errors());
            }else {
                while interface.varients.contains_key(&error_ty) {
                    error_ty = format!("{}-grpc-errors", error_ty);
                }

                interface.varients.insert(error_ty.clone(), grpc_errors());
            }

            Valid::succeed((interface, error_ty))
        })
}
