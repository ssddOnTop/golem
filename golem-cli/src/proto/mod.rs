mod proto;
mod handle_types;
mod handle_services;
mod handle_well_known;

pub static DEFAULT_INTERFACE_NAME: &str = "type";
