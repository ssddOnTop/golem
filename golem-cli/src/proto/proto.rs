use anyhow::anyhow;
use protox::prost_reflect::prost_types::FileDescriptorSet;
use tailcall_valid::{Valid, Validator};
use crate::wit_config::config::WitConfig;
use crate::wit_config::wit_types::WitType;
use crate::proto::handle_services::handle_services;
use crate::proto::handle_types::handle_types;
use crate::proto::handle_well_known::handle_well_known;

pub struct Proto(Vec<FileDescriptorSet>);

impl Proto {
    pub fn new<T: IntoIterator<Item=FileDescriptorSet>>(v: T) -> Self {
        Self(v.into_iter().collect())
    }
    pub fn to_config<T: AsRef<str>>(&self, package_name: T) -> Valid<WitConfig, anyhow::Error, anyhow::Error> {
        Valid::succeed(WitConfig::default())
            .and_then(|config| handle_types(config, &self.0, package_name.as_ref().to_string()))
            .and_then(|config| handle_well_known(config))
            .and_then(|(config, appended_types)| handle_services(config, &self.0, Some(appended_types.appended_error)))
    }
}

pub fn process_ty(name: &str) -> Valid<WitType, anyhow::Error, anyhow::Error> {
    if !name.starts_with('.') {
        return Valid::fail(anyhow!("Expected fully-qualified name for reference type but got {}. This is a bug!", name));
    }
    let name = &name[1..];
    if let Some((_package, name)) = name.rsplit_once('.') {
        Valid::succeed(WitType::FieldTy(name.to_string()))
    }else {
        Valid::succeed(WitType::FieldTy(name.to_string()))
    }
}

#[cfg(test)]
test_r::enable!();

#[cfg(test)]
mod tests {
    use test_r::test;
    use tailcall_valid::Validator;
    use wit_parser::Resolve;
    use crate::proto::proto::Proto;

    #[test]
    fn test_nested() {
        let relative = format!("{}/src/proto/fixtures",env!("CARGO_MANIFEST_DIR"));
       let proto = protox::compile([format!("{}/address.proto", relative)], [relative]).unwrap();
        let proto = Proto::new([proto]);
        let config = proto.to_config("api:todos@1.0.0").to_result().unwrap();

        let mut resolve = Resolve::new();
        assert!(resolve.push_str("foox.wit", &config.to_wit()).is_ok());

        insta::assert_snapshot!(config.to_wit());
    }
}