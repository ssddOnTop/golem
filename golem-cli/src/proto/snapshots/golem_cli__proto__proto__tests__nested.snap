---
source: golem-cli/src/proto/proto.rs
expression: config.to_wit()
---
package api:todos@1.0.0;

interface address-service {
    use %type.{address, get-user-request, user};
    
    
    add-address: func(input: address) -> option<address>;
    get-address-for-user: func(input: user) -> option<address>;
}

interface %type {
    
    record address { city: string, street: string, user: user, zip-code: string }
    record get-user-request { id: string }
    record user { email: string, id: string, name: string }
    
    
}

interface user-service {
    use %type.{address, get-user-request, user};
    
    
    get-user: func(input: get-user-request) -> option<user>;
    create-user: func(input: user) -> option<user>;
}
